docs/question/server.md

```diff
@@ -134,16 +242,97 @@ def generate(data):
     data["params"]["a"] = random.choice([0.25, 0.5, 1, 2, 4])
 
 def file(data):
-    # We should look at data["filename"], generate the corresponding file,
-    # and return the contents of the file as a string, bytes-like, or file-like object.
-    # We can access data["params"].
-    # As an example, we will generate the "fig.png" figure.
-
-    if data["filename"] == "fig.png":                # check for the appropriate filename
-        plt.plot([0, data["params"]["a"]], [0, 1])   # plot a line with slope "a"
-        buf = io.BytesIO()                           # make a bytes object (a buffer)
-        plt.savefig(buf, format="png")               # save the figure data into the buffer
+    # check for the appropriate filename
+    if data["filename"] == "fig.png":
+        # plot a line with slope "a"
+        plt.plot([0, data["params"]["a"]], [0, 1])
+        # make a bytes object (a buffer)
+        buf = io.BytesIO()
+        # save the figure data into the buffer
+        plt.savefig(buf, format="png")
         return buf
 ```
 
-You can also use this functionality in file-based elements (`pl-figure`, `pl-file-download`) by setting `type="dynamic"`.
+Requests for files to this URL will be routed to the `file()` function in `server.py` since the `type` of `pl-figure` is set to `"dynamic"`. The filename requested is stored in `data["filename"]`, and the file contents should be returned from the `file()` function. The URL for the dynamically generated files is set to `{{options.client_files_question_dynamic_url}}` in the HTML, which is a special URL that PrairieLearn uses to route requests to the `file()` function.
```

Comment:
Can we move the nmote about `{{options...}}` to the "advanced usage" section below?

Reply:
Moved in [180b6d8](https://github.com/PrairieLearn/PrairieLearn/pull/11830/commits/180b6d841e4202cfcceadd30f94c48524a62b60c).